### PR TITLE
fix: [DHIS2-15967] allow enrolling TEIs with mandatory TET attributes

### PR DESCRIPTION
**Pull request by superskip**
This fix is temporary and should be reverted when the corresponding backend fix is released.

### CODE REVIEW

**Review by simonadomnisoru — COMMENTED** [review]
Hey @Tony,

I added a suggestion about moving the logic inside the `DataElementFactory`.

In addition to this, I was wondering if it is possible to fix the same bug when editing a TEI in the `WidgetProfile` (because the self-contained `WidgetProfile` has its own logic for creating the form, it won't benefit from the general solution).
One suggestion to fix this is to pass [trackedEntityType.trackedEntityTypeAttributes](https://github.com/dhis2/capture-app/blob/master/src/core_modules/capture-core/components/WidgetProfile/DataEntry/FormFoundation/RenderFoundation.js#L150) down to [setBaseProperties](https://github.com/dhis2/capture-app/blob/master/src/core_modules/capture-core/components/WidgetProfile/DataEntry/FormFoundation/DataElement.js#L104). The current [dataElement.compulsory](https://github.com/dhis2/capture-app/blob/master/src/core_modules/capture-core/components/WidgetProfile/DataEntry/FormFoundation/DataElement.js#L118) logic is incorrect and should be based on both `programTrackedEntityAttribute.mandatory` and the corresponding value inside `trackedEntityType.trackedEntityTypeAttributes`.

Thank you!

File: src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/useMetadataForRegistrationForm/useMetadataForRegistrationForm.js [hunk 1]
```diff
@@ -72,6 +72,20 @@ export const useMetadataForRegistrationForm = ({ selectedScopeId }: Props) => {
...
     });
 
+    if (trackedEntityType && enrollment) {
+        for (const tetAttribute of trackedEntityType.trackedEntityTypeAttributes) {
+            if (tetAttribute.mandatory) {
+                const section = enrollment.enrollmentForm.getSection(Section.MAIN_SECTION_ID);
+                if (section) {
+                    const enrollmentAttribute = section.elements && section.elements.get(tetAttribute.trackedEntityAttributeId);
+                    if (enrollmentAttribute) {
+                        enrollmentAttribute.compulsory = true;
```
[the remark this answers is not in the document]
Comment (superskip) [reply src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/useMetadataForRegistrationForm/useMetadataForRegistrationForm.js:82]: You're right - I did miss a spot with the profile widget! I'll try your suggestion : )

**Comment by superskip** [comment]
@simonadomnisoru  - Refactored the code roughly like you suggested, and patched the profile widget in a similar manner. The fix got more complicated this way, but maybe also a little more robust.

**Review by simonadomnisoru — APPROVED** [review]
Looks great! Thanks for looking into my suggestion

**Review by geethaalwan — APPROVED** [review]
Tested successfully on 2.41,2.40.3,2.39.4,2.38.6 versions

**Pull request merged**